src/toil/realtimeLogger.py

```diff
@@ -0,0 +1,190 @@
+        
+        Messages are 4-byte-length-prefixed JSON-encoded logging module records.
+        """
+        
+        while True:
+            # Loop until we run out of messages
+        
+            # Parse the length
+            length_data = self.rfile.read(4)
+            if len(length_data) < 4:
+                # The connection was closed, or we didn't get enough data
+                # TODO: complain?
+                break
+                
+            # Actually parse the length
+            length = struct.unpack(">L", length_data)[0]
+            
+            # This is where we'll put the received message
+            message_parts = []
+            length_received = 0
+            while length_received < length:
+                # Keep trying to get enough data
+                part = self.rfile.read(length - length_received)
+                
+                length_received += len(part)
+                message_parts.append(part)
+                
+            # Stitch it all together
+            message = "".join(message_parts)
+
+            try:
+            
+                # Parse it as JSON
+                message_attrs = json.loads(message)
+                
+                # Fluff it up into a proper logging record
+                record = logging.makeLogRecord(message_attrs)
+            except:
+                logging.error("Malformed record")
+                
+            # TODO: do log level filtering
+            logging.getLogger("remote").handle(record)
+            
+class JSONDatagramHandler(logging.handlers.DatagramHandler):
+    """
+    Send logging records over UDP serialized as JSON.
+    """
+    
+    def makePickle(self, record):
+        """
+        Actually, encode the record as length-prefixed JSON instead.
+        """
+        
+        json_string = json.dumps(record.__dict__)
+        length = struct.pack(">L", len(json_string))
+        
+        return length + json_string
+        
+class RealtimeLogger(object):
+    """
+    All-static class for getting a logger that logs over UDP to the master.
+    
+    Usage:
+    
+    1. Make sure Job.Runner.startToil() is running on your master.
+    
+    2. From a running job on a worker, do:
+    
+    from toil.realtimeLogger import RealtimeLogger
+    realtime = RealtimeLogger.get()
+    realtime.info("This is a logging message that goes straight to the master")
+    
+    """
+    
+    # Also the logger
+    logger = None
+    
+    # The master keeps a server and thread
+    logging_server = None
+    server_thread = None
+  
+    @classmethod
+    def start_master(cls):
+        """
+        Start up the master server and put its details into the options
+        namespace.
+        
+        Python logging should have already been configured.
+        
+        """
+        
+        # Start up the logging server
+        cls.logging_server = SocketServer.ThreadingUDPServer(("0.0.0.0", 0),
```

Comment:
Ah, I see. Ephemeral port. That should work. Disregards my above comment about configuring a port.